src/model/Snapping.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * Xournal++
+ *
+ * Snapping methods
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <cmath>
+
+/**
+ * @brief A class providing methods for snapping to grid and rotation snapping.
+ */
+
+class Snapping {
+public:
+    /**
+     * @brief Replaces a point by the nearest grid point, if its distance therefrom is under
+     * a certain tolerance
+     * @param x the x-coordinate of the point
+     * @param y the y-coordinate of the point
+     * @param t the tolerance
+     */
+    static void snapToGrid(double& x, double& y, double t);
+
+    /**
+     * @brief Replaces an angle close to a multiple of M_PI / 4 by the closest multiple of M_PI / 4
+     * @param a the angle (assumed to be between - M_PI and M_PI)
+     */
+    static void snapAngle(double& a, double t);
+
+    /**
+     * @brief Snaps the angle between the horizontal axis and the line between the given point and center
+     * and replaces the point with the point rotated accordingly
+     * @param x the x-coordinate of the point
+     * @param y the y-coordinate of the point
+     * @param cx the x-coordinate of the center of rotation
+     * @param cy the y-coordinate of the center of rotation
+     * @param t the tolerance
+     */
+    static void snapRotation(double& x, double& y, double cx, double cy, double t);
+};
```

Comment:
Thanks, Febbe, for all suggestions and explanations. One question: With your suggested version of Snapping.h we will have two implementations of a point. The simple one as a struct from Snapping.h and the one from Point.h. Why not to use the latter one? Is this simply a question of performance? It must be said that the simple points must also be casted to those from Point.h if they are to be used in the tool handlers. So is this really worth it to have these two definitions of a point?

Reply:
It does not matter if we give the coordinates as separate parameters x and y or as "parameter pack" in c and c++ this is logically the same. Also it's rather a point of coherence instead of performance. Both x and y are part of a "Point" or "Coordinate" and belong strictly together. Also it increases readability without writing / reading the documentation. And it allows us to return 2 parameters at once per value.
I didn't used the Point from the model (MP) because it contained 2 extra parameters, which are never used by those functions. I didn't cared much about performance, because the compiler could optimize those extra parameters away if it inlines the function.
You could also just use the MP, but then the functions are limited to those MPs. Other developers need to either rewrite everything or use the function by creating a MP in order to use them. I bet they will do the latter ;).

I propose 3 ways for you:
 1. use the header from me as is and assign x and y, maybe in a helper function to MP.
 2. use the header from me as is and extract the simple Point, rename MP to something more meaningful and make the simple Point part of the renamed MP, then assign the return value to the point parameter of the renamed MP
 3. use the MP as `Point const& pos` parameter but be aware, you need to copy the whole point before returning it: 
```
Point snapToGrid(Point const& pos, double gridSize, double tolerance) {
    Point ret{pos};
    [calculate and assign x, y ... ]
```  

I preffere the second, but you'll have to do work a maintainer should do.

Reply:
I chose the third way. It seems simpler to me of just reusing MPs, even if the pressure is used nowhere. The same applies to other parts of the xournal++ code, too.

Reply:
I will take a look later this day or tomorrow evening.